### ISSUE TITLE
Configuration ‘switchInterval’ is not honored for the first switch.

### ISSUE THREAD

**Issue by mraavi** [google/shaka-player#2377]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
YES

**Is your feature request related to a problem? Please describe.**

I tried the Shaka configuration ‘switchInterval’ set to 18 secs (and greater values) but Shaka not honoring this configuration for the first switch. 
I dug in to Shaka code 'simple_abr_manager' to figure out the reason. From the code it looks like, for the first switch shaka player don’t consider this configuration. This configuration applied to only subsequent switches. For the first switch, shaka try to adapt to available bandwidth as soon as possible. 
The issue i am seeing on Samsung TV's is media starts only after 3 video segment downloads (Filed a separate issue: https://github.com/google/shaka-player/issues/2291). So i want to restrict bitrate switching until media starts (18 sec). But, irrespective of 'switchInterval' value bitrate switch always happens after 2nd segment download (I assume, to calculate bandwidth availability it is taking two segments)
Can we make 'switchInterval' configuration applies even for first bitrate switch?

**Describe the solution you'd like**

Can we make 'switchInterval' configuration applies even for first bitrate switch?

**Comment by joeyparrish** [comment]
Here's what I'm seeing in the latest code from the `master` branch.  (You didn't mention what version of Shaka player you're using.  Next time, please fill out our issue template in full.)

All decisions are made in `chooseVariant()`.  When that happens, we note the time in `this.lastTimeChosenMs_`.

There are three ways to call into `chooseVariant()`:
  1. When we choose the initial streams during load.  This is not considered a switch.
  2. When we get our first meaningful bandwidth estimate.  This is considered a switch.
  3. When it has been more than `this.config_.switchInterval_` seconds since our last switch.

Switching as soon as we have a meaningful bandwidth estimate is intended behavior.  As soon as we know what is ideal for the user, we want to start streaming that.

What I don't understand is why you want to change this.  You said:

> Samsung TV's is media starts only after 3 video segment downloads. So i want to restrict bitrate switching until media starts (18 sec).

I don't understand the relationship between these things.  Why would restricting bitrate switching help?  By default, we don't clear anything from the buffer when we switch bitrates.  If we start with stream 0 and then switch to stream 1 between the second & third segment, your buffer would be:

[Stream 0, segment 1] [Stream 0, segment 2] [Stream **0**, segment 3] [Stream **1**, segment 4]

**Comment by joeyparrish** [comment]
Also, can you please provide your Shaka Player version and the complete configuration you're applying to it?  It's fine to censor license server URLs and such, but we need to see any non-standard config values you're using to completely assess the situation.

**Comment by mraavi** [comment]
> Also, can you please provide your Shaka Player version and the complete configuration you're applying to it? It's fine to censor license server URLs and such, but we need to see any non-standard config values you're using to completely assess the situation.

We are using Shaka player version v2.5.3.
Configuration applied,
       ```
 mPlayer.configure('manifest.dash.ignoreMinBufferTime', true);
mPlayer.configure('manifest.dash.clockSyncUri', headerUrl);
        mPlayer.configure({
            abr: {
                defaultBandwidthEstimate: 1000000,
                switchInterval: 30
            },
            streaming: {
                bufferBehind: 12,
                bufferingGoal: 30,
                rebufferingGoal: 1
            }
        });
```

**Comment by mraavi** [comment]
> Here's what I'm seeing in the latest code from the `master` branch. (You didn't mention what version of Shaka player you're using. Next time, please fill out our issue template in full.)
> 
> All decisions are made in `chooseVariant()`. When that happens, we note the time in `this.lastTimeChosenMs_`.
> 
> There are three ways to call into `chooseVariant()`:
> 
>     1. When we choose the initial streams during load.  This is not considered a switch.
> 
>     2. When we get our first meaningful bandwidth estimate.  This is considered a switch.
> 
>     3. When it has been more than `this.config_.switchInterval_` seconds since our last switch.
> 
> 
> Switching as soon as we have a meaningful bandwidth estimate is intended behavior. As soon as we know what is ideal for the user, we want to start streaming that.
> 
> What I don't understand is why you want to change this. You said:
> 
> > Samsung TV's is media starts only after 3 video segment downloads. So i want to restrict bitrate switching until media starts (18 sec).
> 
> I don't understand the relationship between these things. Why would restricting bitrate switching help? By default, we don't clear anything from the buffer when we switch bitrates. If we start with stream 0 and then switch to stream 1 between the second & third segment, your buffer would be:
> 
> [Stream 0, segment 1] [Stream 0, segment 2] [Stream **0**, segment 3] [Stream **1**, segment 4]

We are trying to download and start media with lowest media variant so that we can start the media faster. Ex. Downloading lowest variant takes about 0.5 sec where as downloading highest media variant takes about 1.5 sec .
Because switch happening as soon as we have a meaningful bandwidth estimate, highest variant is downloaded as last segment. This is impacting media start up time, because media starts only after 3 segment downloads.
1. If media can start after 1st segment download, this (switch happening as soon as we have a meaningful bandwidth ) should not be an issue.
2. But because media is starting only after 3 media segments on SamsungTV's, we wanted to limit first 3 segments to lowest variant.

**Comment by joeyparrish** [comment]
We can't reproduce your results about the media only starting after 3 segments.  At least, not in the latest release, v2.5.9.  Our observation in #2291 was that the only way for playback to start after more than 1 segment was if the segments were all fetched very, very rapidly.

It seems that switchInterval is behaving as it is designed to, and that you only need this to work around issue #2291.  If you'd like to modify the behavior of switchInterval, you are welcome to do so in your own fork and your own deployment of Shaka Player.  I don't believe this is something we want to change at this time.

Instead, we will focus our effort on helping you resolve #2291, which is the root issue you're having.
